File: packages/sdks/client-sdks/gatrix-js-client-sdk/src/VariationProvider.ts
/**
 * VariationProvider - Interface for centralized variation logic.
 *
 * This interface defines the internal methods that handle all variation logic:
 * flag lookup, value extraction, and metrics tracking.
 *
 * FeaturesClient implements this interface.
 * FlagProxy holds a reference to this interface (not FeaturesClient directly)
 * to avoid circular dependencies and keep a clean separation.
 *
 * All methods accept an optional `forceRealtime` parameter (default: false).
 * When true, the method reads from realtimeFlags directly, bypassing
 * explicitSyncMode's synchronizedFlags.
 */
import { EvaluatedFlag, Variant, ValueType, VariationResult } from './types';

export interface VariationProvider {
  // Core access
  isEnabledInternal(flagName: string, forceRealtime?: boolean): boolean;
  getVariantInternal(flagName: string, forceRealtime?: boolean): Variant;

  // Metadata access (no metrics tracking)
  hasFlagInternal(flagName: string, forceRealtime?: boolean): boolean;
  getValueTypeInternal(flagName: string, forceRealtime?: boolean): ValueType;
  getVersionInternal(flagName: string, forceRealtime?: boolean): number;
  getReasonInternal(flagName: string, forceRealtime?: boolean): string | undefined;
  getImpressionDataInternal(flagName: string, forceRealtime?: boolean): boolean;
  getRawFlagInternal(flagName: string, forceRealtime?: boolean): EvaluatedFlag | undefined;

  // Variation (returns value)
  variationInternal(flagName: string, fallbackValue: string, forceRealtime?: boolean): string;
  boolVariationInternal(flagName: string, fallbackValue: boolean, forceRealtime?: boolean): boolean;
  stringVariationInternal(flagName: string, fallbackValue: string, forceRealtime?: boolean): string;
  numberVariationInternal(flagName: string, fallbackValue: number, forceRealtime?: boolean): number;
  jsonVariationInternal<T>(flagName: string, fallbackValue: T, forceRealtime?: boolean): T;

  // Variation Details (returns value + reason)
  boolVariationDetailsInternal(
    flagName: string,
    fallbackValue: boolean,
    forceRealtime?: boolean
  ): VariationResult<boolean>;
  stringVariationDetailsInternal(
    flagName: string,
    fallbackValue: string,
    forceRealtime?: boolean
  ): VariationResult<string>;
  numberVariationDetailsInternal(
    flagName: string,
    fallbackValue: number,
    forceRealtime?: boolean
  ): VariationResult<number>;
  jsonVariationDetailsInternal<T>(
    flagName: string,
    fallbackValue: T,
    forceRealtime?: boolean
  ): VariationResult<T>;

  // Strict Variation (throws on missing/mismatch)
  boolVariationOrThrowInternal(flagName: string, forceRealtime?: boolean): boolean;
  stringVariationOrThrowInternal(flagName: string, forceRealtime?: boolean): string;
  numberVariationOrThrowInternal(flagName: string, forceRealtime?: boolean): number;
  jsonVariationOrThrowInternal<T>(flagName: string, forceRealtime?: boolean): T;
}

File: packages/sdks/client-sdks/gatrix-flutter-client-sdk/lib/src/variation_provider.dart
import 'models.dart';

/// Interface for variation logic delegation.
/// FlagProxy delegates variation calls to this interface,
/// which is implemented by FeaturesClient.
/// This prevents circular dependencies between FlagProxy and FeaturesClient.
///
/// All methods accept an optional [forceRealtime] parameter (default: false).
/// When true, reads from realtimeFlags directly, bypassing explicitSyncMode.
abstract class VariationProvider {
  bool isEnabledInternal(String flagName, {bool forceRealtime = true});
  Variant getVariantInternal(String flagName, {bool forceRealtime = true});

  // Metadata access (no metrics tracking)
  bool hasFlagInternal(String flagName, {bool forceRealtime = true});
  ValueType getValueTypeInternal(String flagName, {bool forceRealtime = true});
  int getVersionInternal(String flagName, {bool forceRealtime = true});
  String? getReasonInternal(String flagName, {bool forceRealtime = true});
  bool getImpressionDataInternal(String flagName, {bool forceRealtime = true});
  EvaluatedFlag? getRawFlagInternal(String flagName, {bool forceRealtime = true});
  String variationInternal(String flagName, String fallbackValue,
      {bool forceRealtime = true});
  bool boolVariationInternal(String flagName, bool fallbackValue,
      {bool forceRealtime = true});
  String stringVariationInternal(String flagName, String fallbackValue,
      {bool forceRealtime = true});
  int intVariationInternal(String flagName, int fallbackValue,
      {bool forceRealtime = true});
  double doubleVariationInternal(String flagName, double fallbackValue,
      {bool forceRealtime = true});
  T jsonVariationInternal<T>(String flagName, T fallbackValue,
      {bool forceRealtime = true});

  VariationResult<bool> boolVariationDetailsInternal(
      String flagName, bool fallbackValue,
      {bool forceRealtime = true});
  VariationResult<String> stringVariationDetailsInternal(
      String flagName, String fallbackValue,
      {bool forceRealtime = true});
  VariationResult<int> intVariationDetailsInternal(
      String flagName, int fallbackValue,
      {bool forceRealtime = true});
  VariationResult<double> doubleVariationDetailsInternal(
      String flagName, double fallbackValue,
      {bool forceRealtime = true});
  VariationResult<T> jsonVariationDetailsInternal<T>(
      String flagName, T fallbackValue,
      {bool forceRealtime = true});

  bool boolVariationOrThrowInternal(String flagName, {bool forceRealtime = true});
  String stringVariationOrThrowInternal(String flagName,
      {bool forceRealtime = true});
  int intVariationOrThrowInternal(String flagName, {bool forceRealtime = true});
  double doubleVariationOrThrowInternal(String flagName,
      {bool forceRealtime = true});
  T jsonVariationOrThrowInternal<T>(String flagName, {bool forceRealtime = true});
}
